modules/packages/Futures.chpl

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+Containers for accessing the results of asynchronous execution.
+
+A :record:`Future` object is a container that can store the result of an
+asynchronous operation, which can be retrieved when the result is ready.
+
+Usage
+-----
+
+A :record:`Future` object is not created directly. Instead, a future may be
+created by calling the :proc:`async()` function, which takes as arguments
+the function to be executed and all arguments to that function.
+
+The following example demonstrates a trivial use of futures. Three computations
+are executed asynchronously.
+
+.. code-block:: chapel
+
+   use Futures;
+
+   config const X = 42;
+
+   const A = async(lambda(x: int) { return 2 * x; }, X);
+   const B = async(lambda(x: int) { return 3 * x; }, X);
+   const C = async(lambda(x: int) { return 4 * x; }, X);
+
+   writeln(A.get());
+   writeln(B.get());
+   writeln(C.get());
+
+Future Chaining
+---------------
+
+A continuation to a future (itself a future) can be created via the
+:proc:`Future.andThen()` method, which takes as its single argument a function
+to be invoked asynchronously (with respect to other tasks) but strictly ordered
+in execution after the result of the parent future is ready. The continuation
+function takes a single argument, the result of the parent future.
+
+The following examples demonstrate such chaining of futures.
+
+.. code-block:: chapel
+
+   use Futures;
+
+   config const X = 42;
+
+   const F = async(lambda(x: int) { return 2 * x; }, X)
+     .andThen(lambda(x: int) { return x + 7; })
+     .andThen(lambda(x: int) { return 3 * x; });
+
+   writeln(F.get() == (3 * ((2 * X) + 7))); // prints "true"
+
+.. code-block:: chapel
+
+   use Futures;
+
+   config const X = 1234;
+
+   var F = async(lambda(x:int) { return x:string; }, X)
+     .andThen(lambda(x:string) { return(x.length); });
+
+   writeln(F.get()); // prints "4"
+
+ */
+
+module Futures {
+
+  use Reflection;
+  use RefCount;
+
+  pragma "no doc"
+  class FutureClass: RefCountBase {
```

Comment:
I think Chapel has a much higher likelihood for code to be implicitly parallel than does C++. Assuming that, I think shared-by-default is helpful and more intuitive to users. If the future isn't shared between tasks, I wouldn't expect reference counting to incur too much overhead, since I doubt users would create many handles to the same future.